File: TPs/TP02/Insercao.java
// Algoritmos e Estruturas de Dados 2 ----- Trabalho Prático 02
// Questão 4 ----- Ordenação por Inserção em Java

import java.io.*;
import java.util.Scanner;

public class Insercao {

    static long comparacoes = 0;
    static long movimentacoes = 0;

    public static void main(String[] args) throws Exception {

        long inicio = System.nanoTime();

        Scanner sc = new Scanner(System.in);
        ColecaoRestaurantes colecao = ColecaoRestaurantes.lerCsv();

        Restaurante[] selecionados = new Restaurante[1000];
        int n = 0;

        int id;
        while ((id = sc.nextInt()) != -1) {

            for (int i = 0; i < colecao.getTamanho(); i++) {
                if (colecao.getRestaurantes()[i].getId() == id) {
                    selecionados[n++] = colecao.getRestaurantes()[i];
                    break;
                }
            }
        }

        insercao(selecionados, n);

        for (int i = 0; i < n; i++) {
            System.out.println(selecionados[i].formatar());
        }

        long fim = System.nanoTime();
        double tempo = (fim - inicio) / 1e9;

        FileWriter fw = new FileWriter("matricula_insercao.txt");
        fw.write("123456\t" + comparacoes + "\t" + movimentacoes + "\t" + tempo);
        fw.close();

        sc.close();
    }

    // Inserção
    public static void insercao(Restaurante[] array, int n) {

        for (int i = 1; i < n; i++) {

            Restaurante tmp = array[i];
            movimentacoes++;

            int j = i - 1;

            while (j >= 0) {
                comparacoes++;

                if (array[j].getCidade().compareTo(tmp.getCidade()) > 0) {
                    array[j + 1] = array[j];
                    movimentacoes++;
                    j--;
                } else {
                    break;
                }
            }

            array[j + 1] = tmp;
            movimentacoes++;
        }
    }
}

class Data {
    private int dia, mes, ano;

    public Data(int d, int m, int a) {
        dia = d; mes = m; ano = a;
    }

    public static Data parseData(String s) {
        int ano = Integer.parseInt(s.substring(0, 4));
        int mes = Integer.parseInt(s.substring(5, 7));
        int dia = Integer.parseInt(s.substring(8, 10));
        return new Data(dia, mes, ano);
    }

    public String formatar() {
        return String.format("%02d/%02d/%04d", dia, mes, ano);
    }
}

class Hora {
    private int hora, minuto;

    public Hora(int h, int m) {
        hora = h; minuto = m;
    }

    public static Hora parseHora(String s) {
        int h = Integer.parseInt(s.substring(0, 2));
        int m = Integer.parseInt(s.substring(3, 5));
        return new Hora(h, m);
    }

    public String formatar() {
        return String.format("%02d:%02d", hora, minuto);
    }
}

class Restaurante {

    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tipos;
    private int faixaPreco;
    private Hora abertura, fechamento;
    private Data data;
    private boolean aberto;

    public int getId() { return id; }
    public String getCidade() { return cidade; }

    public static Restaurante parseRestaurante(String linha) {

        String[] c = linha.split(",");

        Restaurante r = new Restaurante();

        r.id = Integer.parseInt(c[0]);
        r.nome = c[1];
        r.cidade = c[2];
        r.capacidade = Integer.parseInt(c[3]);
        r.avaliacao = Double.parseDouble(c[4]);

        r.tipos = c[5].split(";");
        r.faixaPreco = c[6].length();

        String[] h = c[7].split("-");
        r.abertura = Hora.parseHora(h[0]);
        r.fechamento = Hora.parseHora(h[1]);

        r.data = Data.parseData(c[8]);
        r.aberto = c[9].equals("true");

        return r;
    }

    public String formatar() {

        String tiposStr = "[";
        for (int i = 0; i < tipos.length; i++) {
            tiposStr += tipos[i];
            if (i < tipos.length - 1) tiposStr += ",";
        }
        tiposStr += "]";

        String preco = "";
        for (int i = 0; i < faixaPreco; i++) preco += "$";

        return "["
                + id + " ## "
                + nome + " ## "
                + cidade + " ## "
                + capacidade + " ## "
                + avaliacao + " ## "
                + tiposStr + " ## "
                + preco + " ## "
                + abertura.formatar() + "-" + fechamento.formatar() + " ## "
                + data.formatar() + " ## "
                + aberto
                + "]";
    }
}

class ColecaoRestaurantes {

    private int tamanho;
    private Restaurante[] array;

    public int getTamanho() { return tamanho; }
    public Restaurante[] getRestaurantes() { return array; }

    public void lerCsv(String path) throws Exception {

        BufferedReader br = new BufferedReader(new FileReader(path));

        array = new Restaurante[10000];
        tamanho = 0;

        br.readLine();

        String linha;
        while ((linha = br.readLine()) != null) {
            array[tamanho++] = Restaurante.parseRestaurante(linha);
        }

        br.close();
    }

    public static ColecaoRestaurantes lerCsv() throws Exception {
        ColecaoRestaurantes c = new ColecaoRestaurantes();
        c.lerCsv("/tmp/restaurantes.csv");
        return c;
    }
}

File: TPs/TP03/Insercao.c
//Algoritmos e Estruturas de Dados 2 ----- Trabalho Prático 3
//Questão 02 ----- Ordenação PARCIAL por Inserção em C

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct Data {
    int ano;
    int mes;
    int dia;
} Data;

typedef struct Hora {
    int hora;
    int minuto;
} Hora;

typedef struct Restaurante {
    int id;
    char nome[200];
    char cidade[100];
    int capacidade;
    double avaliacao;

    char tipos_cozinha[20][100];
    int quantidade_tipos;

    int faixa_preco;

    Hora horario_abertura;
    Hora horario_fechamento;

    Data data_abertura;

    int aberto;
} Restaurante;

typedef struct ColecaoRestaurantes {
    Restaurante* restaurantes[10000];
    int tamanho;
} ColecaoRestaurantes;

long comparacoes = 0;
long movimentacoes = 0;

Data parse_data(char* s) {

    Data data;

    sscanf(s, "%d-%d-%d",
           &data.ano,
           &data.mes,
           &data.dia);

    return data;
}

Hora parse_hora(char* s) {

    Hora hora;

    sscanf(s, "%d:%d",
           &hora.hora,
           &hora.minuto);

    return hora;
}

void formatar_data(Data* data, char* buffer) {

    sprintf(buffer,
            "%02d/%02d/%04d",
            data->dia,
            data->mes,
            data->ano);
}

void formatar_hora(Hora* hora, char* buffer) {

    sprintf(buffer,
            "%02d:%02d",
            hora->hora,
            hora->minuto);
}

void separar_csv(char* linha, char campos[][300]) {

    int i = 0;
    int j = 0;
    int campo = 0;
    int aspas = 0;

    while (linha[i] != '\0' && linha[i] != '\n') {

        if (linha[i] == '"') {
            aspas = !aspas;
        }
        else if (linha[i] == ',' && aspas == 0) {

            campos[campo][j] = '\0';

            campo++;
            j = 0;
        }
        else {

            campos[campo][j] = linha[i];
            j++;
        }

        i++;
    }

    campos[campo][j] = '\0';
}

Restaurante* parse_restaurante(char* linha) {

    Restaurante* restaurante =
        (Restaurante*) malloc(sizeof(Restaurante));

    char campos[10][300];

    separar_csv(linha, campos);

    restaurante->id = atoi(campos[0]);

    strcpy(restaurante->nome, campos[1]);

    strcpy(restaurante->cidade, campos[2]);

    restaurante->capacidade = atoi(campos[3]);

    restaurante->avaliacao = atof(campos[4]);

    restaurante->quantidade_tipos = 0;

    int i = 0;
    int j = 0;
    int tipo = 0;

    while (campos[5][i] != '\0') {

        if (campos[5][i] == ';') {

            restaurante->tipos_cozinha[tipo][j] = '\0';

            tipo++;
            j = 0;
        }
        else {

            restaurante->tipos_cozinha[tipo][j] =
                campos[5][i];

            j++;
        }

        i++;
    }

    restaurante->tipos_cozinha[tipo][j] = '\0';

    restaurante->quantidade_tipos = tipo + 1;

    restaurante->faixa_preco = strlen(campos[6]);

    char horario1[20];
    char horario2[20];

    sscanf(campos[7],
           "%[^-]-%s",
           horario1,
           horario2);

    restaurante->horario_abertura =
        parse_hora(horario1);

    restaurante->horario_fechamento =
        parse_hora(horario2);

    restaurante->data_abertura =
        parse_data(campos[8]);

    restaurante->aberto =
        strcmp(campos[9], "true") == 0;

    return restaurante;
}

void formatar_restaurante(Restaurante* restaurante,
                          char* buffer) {

    char data[30];

    char abertura[20];
    char fechamento[20];

    formatar_data(&restaurante->data_abertura,
                  data);

    formatar_hora(&restaurante->horario_abertura,
                  abertura);

    formatar_hora(&restaurante->horario_fechamento,
                  fechamento);

    char tipos[500] = "[";

    for (int i = 0;
         i < restaurante->quantidade_tipos;
         i++) {

        strcat(tipos,
               restaurante->tipos_cozinha[i]);

        if (i <
            restaurante->quantidade_tipos - 1) {

            strcat(tipos, ",");
        }
    }

    strcat(tipos, "]");

    char preco[10] = "";

    for (int i = 0;
         i < restaurante->faixa_preco;
         i++) {

        strcat(preco, "$");
    }

    sprintf(buffer,
            "[%d ## %s ## %s ## %d ## %.1lf ## %s ## %s ## %s-%s ## %s ## %s]",
            restaurante->id,
            restaurante->nome,
            restaurante->cidade,
            restaurante->capacidade,
            restaurante->avaliacao,
            tipos,
            preco,
            abertura,
            fechamento,
            data,
            restaurante->aberto ? "true" : "false");
}

void ler_csv_colecao(ColecaoRestaurantes* colecao,
                     char* path) {

    FILE* arquivo = fopen(path, "r");

    char linha[1000];

    fgets(linha, 1000, arquivo);

    colecao->tamanho = 0;

    while (fgets(linha, 1000, arquivo) != NULL) {

        colecao->restaurantes[colecao->tamanho] =
            parse_restaurante(linha);

        colecao->tamanho++;
    }

    fclose(arquivo);
}

Restaurante* buscar_por_id(ColecaoRestaurantes* colecao,
                           int id) {

    Restaurante* resp = NULL;

    for (int i = 0;
         i < colecao->tamanho;
         i++) {

        if (colecao->restaurantes[i]->id == id) {

            resp = colecao->restaurantes[i];

            i = colecao->tamanho;
        }
    }

    return resp;
}

int comparar(Restaurante* a,
             Restaurante* b) {

    comparacoes++;

    return strcmp(a->cidade,
                  b->cidade);
}

void insercao_parcial(Restaurante* array[],
                      int n,
                      int k) {

    for (int i = 1; i < n; i++) {

        Restaurante* tmp = array[i];

        movimentacoes++;

        int j =
            (i < k) ? i - 1 : k - 1;

        while (j >= 0 &&
               comparar(array[j], tmp) > 0) {

            if (j + 1 < k) {

                array[j + 1] = array[j];

                movimentacoes++;
            }

            j--;
        }

        if (j + 1 < k) {

            array[j + 1] = tmp;

            movimentacoes++;
        }
    }
}

void gerar_log(double tempo) {

    FILE* log =
        fopen("810688_insercao_parcial.txt",
              "w");

    fprintf(log,
            "810688\t%ld\t%ld\t%lf",
            comparacoes,
            movimentacoes,
            tempo);

    fclose(log);
}

int main() {

    clock_t inicio = clock();

    ColecaoRestaurantes colecao;

    ler_csv_colecao(&colecao,
                    "/tmp/restaurantes.csv");

    Restaurante* array[1000];

    int n = 0;

    int id;

    scanf("%d", &id);

    while (id != -1) {

        array[n] =
            buscar_por_id(&colecao, id);

        n++;

        scanf("%d", &id);
    }

    int k = 10;

    if (k > n) {
        k = n;
    }

    insercao_parcial(array, n, k);

    char buffer[1000];

    for (int i = 0; i < n; i++) {

        formatar_restaurante(array[i],
                             buffer);

        printf("%s\n", buffer);
    }

    clock_t fim = clock();

    double tempo =
        ((double)(fim - inicio)) /
        CLOCKS_PER_SEC;

    gerar_log(tempo);

    return 0;
}
